File: jclasslib/src/org/gjt/jclasslib/browser/config/classpath/ClasspathBrowser.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.config.classpath;

import org.gjt.jclasslib.browser.BrowserMDIFrame;
import org.gjt.jclasslib.util.GUIHelper;
import org.gjt.jclasslib.util.ProgressDialog;

import javax.swing.*;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;
import java.awt.*;
import java.awt.event.*;

/**
    Classpath browser that shows a tree of the contents of a
    <tt>ClasspathComponent</tt>.

    @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
    @version $Revision: 1.1 $ $Date: 2003-08-18 08:10:15 $
*/
public class ClasspathBrowser extends JDialog
                              implements ActionListener, ClasspathChangeListener
{

    private static final int DIALOG_WIDTH = 450;
    private static final int DIALOG_HEIGHT = 450;

    private BrowserMDIFrame frame;
    private ClasspathComponent classpathComponent;

    private JLabel lblTitle;
    private JTree tree;
    private JScrollPane scpTree;
    private JButton btnSetup;
    private JButton btnSync;
    private JButton btnOk;
    private JButton btnCancel;

    private ProgressDialog progressDialog;
    private boolean resetOnNextMerge;
    private boolean needsMerge;

    private String selectedClassName;

    /**
     * Constructor.
     * @param frame the parent frame.
     * @param classpathComponent the classpath component to display initially.
     * @param title the disalog title.
     * @param setupVisible if the <i>setup classpath</i> button should be visible.
     */
    public ClasspathBrowser(BrowserMDIFrame frame, ClasspathComponent classpathComponent, String title, boolean setupVisible) {
        super(frame);
        this.frame = frame;

        setClasspathComponent(classpathComponent);
        setupControls(title, setupVisible);
        setupComponent();
        setupEventHandlers();
    }

    public void actionPerformed(ActionEvent event) {
        Object source = event.getSource();
        if (source == btnCancel) {
            doCancel();
        } else if (source == btnOk) {
            doOk();
        } else if (source == btnSetup) {
            doSetup();
        } else if (source == btnSync) {
            doSync(true);
        }

    }

    public void classpathChanged(ClasspathChangeEvent event) {
        needsMerge = true;
        if (event.isRemoval()) {
            resetOnNextMerge = true;
        }
    }

    public void setVisible(boolean visible) {
        if (visible) {
            selectedClassName = null;
        }
        super.setVisible(visible);
    }

    /**
     * Get the name of the selected class.
     * @return the name
     */
    public String getSelectedClassName() {
        return selectedClassName;
    }

    /**
     * Set the new classpath component to be displayed by this dialog.
     * The previous content will be cleared.
     * @param classpathComponent the new classpath component.
     */
    public void setClasspathComponent(ClasspathComponent classpathComponent) {
        if (this.classpathComponent != null) {
            this.classpathComponent.removeClasspathChangeListener(this);
        }
        this.classpathComponent = classpathComponent;
        if (classpathComponent != null) {
            classpathComponent.addClasspathChangeListener(this);
        }
        resetOnNextMerge = true;
        needsMerge = true;
        clear();
    }

    /**
     * Clear the current contents of the dialog. The tree will not be synchronized
     * automatically on the next <tt>setVisible</tt>.
     */
    public void clear() {
        if (tree != null) {
            tree.setModel(new DefaultTreeModel(new ClassTreeNode()));
        }
    }

    private void setupControls(String title, boolean setupVisible) {

        lblTitle = new JLabel(title);
        tree = new JTree(new ClassTreeNode());
        tree.setRootVisible(false);
        tree.setShowsRootHandles(true);
        tree.putClientProperty("JTree.lineStyle", "Angled");
        scpTree = new JScrollPane(tree);

        btnSetup = new JButton("Setup classpath");
        btnSetup.setVisible(setupVisible);
        btnSync = new JButton("Synchronize");
        btnOk = new JButton("Ok");
        btnOk.setEnabled(false);
        btnCancel = new JButton("Cancel");
        btnOk.setPreferredSize(btnCancel.getPreferredSize());

        progressDialog = new ProgressDialog(this, null, "Scanning classpath ...");

    }

    private void setupComponent() {

        Container contentPane = getContentPane();
        contentPane.setLayout(new GridBagLayout());
        GridBagConstraints gc = new GridBagConstraints();
        gc.gridx = 0;
        gc.gridy = 0;
        gc.insets = new Insets(5, 5, 0, 5);
        gc.weightx = 1;
        gc.anchor = GridBagConstraints.NORTHWEST;
        contentPane.add(lblTitle, gc);
        gc.gridy++;

        gc.weighty = 1;
        gc.insets.top = 0;
        gc.fill = GridBagConstraints.BOTH;
        contentPane.add(scpTree, gc);
        gc.gridy++;
        gc.fill = GridBagConstraints.HORIZONTAL;
        gc.weighty = 0;
        gc.insets.top = 3;
        gc.insets.bottom = 5;
        contentPane.add(createButtonBox(), gc);
        getRootPane().setDefaultButton(btnOk);

        setSize(DIALOG_WIDTH, DIALOG_HEIGHT);
        setModal(true);
        setTitle("Choose a class");
        GUIHelper.centerOnParentWindow(this, getOwner());
        setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);

    }

    private Box createButtonBox() {

        Box box = Box.createHorizontalBox();
        box.add(btnSetup);
        box.add(btnSync);
        box.add(Box.createHorizontalGlue());
        box.add(btnOk);
        box.add(btnCancel);

        return box;
    }

    private void setupEventHandlers() {

        btnCancel.addActionListener(this);
        btnOk.addActionListener(this);
        btnSetup.addActionListener(this);
        btnSync.addActionListener(this);

        addWindowListener(new WindowAdapter() {
            public void windowClosing(WindowEvent event) {
                doCancel();
            }
        });
        KeyStroke keyStroke = KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0);
        Object key = new Object();

        JComponent contentPane = (JComponent)getContentPane();
        contentPane.getInputMap(JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT).put(keyStroke, key);
        contentPane.getActionMap().put(key, new AbstractAction() {
            public void actionPerformed(ActionEvent event) {
                doCancel();
            }
        });

        addComponentListener(new ComponentAdapter() {
            public void componentShown(ComponentEvent event) {
                conditionalUpdate();
            }
        });

        tree.addTreeSelectionListener(new TreeSelectionListener() {
            public void valueChanged(TreeSelectionEvent event) {
                checkTreeSelection();
            }
        });

        tree.addMouseListener(new MouseAdapter() {
            public void mouseClicked(MouseEvent event) {
                if (event.getClickCount() == 2 && isValidDoubleClickPath(event)) {
                    doOk();
                }
            }
        });

    }

    private void conditionalUpdate() {
        if (needsMerge) {
            doSync(resetOnNextMerge);
        }
    }

    private boolean isValidDoubleClickPath(MouseEvent event) {

        TreePath locationPath = tree.getPathForLocation(event.getX(), event.getY());
        TreePath selectionPath = tree.getSelectionPath();
        if (selectionPath == null || locationPath == null || !selectionPath.equals(locationPath)) {
            return false;
        }
        ClassTreeNode lastPathComponent = (ClassTreeNode)selectionPath.getLastPathComponent();
        return !lastPathComponent.isPackageNode();
    }

    private void checkTreeSelection() {

        TreePath selectionPath = tree.getSelectionPath();
        boolean enabled = false;
        if (selectionPath != null) {
            ClassTreeNode classTreeNode = (ClassTreeNode)selectionPath.getLastPathComponent();
            enabled = !classTreeNode.isPackageNode();
        }
        btnOk.setEnabled(enabled);
    }

    private void doOk() {

        StringBuffer buffer = new StringBuffer();
        TreePath selectionPath = tree.getSelectionPath();
        for (int i = 1; i < selectionPath.getPathCount(); i++) {
            if (buffer.length() > 0) {
                buffer.append('/');
            }
            buffer.append(selectionPath.getPathComponent(i).toString());
        }
        selectedClassName = buffer.toString();

        setVisible(false);
    }


    private void doCancel() {
        setVisible(false);
    }

    private void doSetup() {
        frame.getActionSetupClasspath().actionPerformed(new ActionEvent(this, 0, null));
        conditionalUpdate();
    }

    private void doSync(final boolean reset) {

        final DefaultTreeModel model = reset ? new DefaultTreeModel(new ClassTreeNode()) : (DefaultTreeModel)tree.getModel();
        Runnable mergeTask = new Runnable() {
            public void run() {
                if (classpathComponent != null) {
                    classpathComponent.mergeClassesIntoTree(model, reset);
                }
            }
        };

        progressDialog.setRunnable(mergeTask);
        progressDialog.setVisible(true);

        if (reset) {
            tree.setModel(model);
        }
        tree.expandPath(new TreePath(model.getRoot()));
        resetOnNextMerge = false;
        needsMerge = false;
    }


}

File: modules/browser/src/main/kotlin/org/gjt/jclasslib/browser/config/classpath/ClasspathBrowser.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.config.classpath

import net.miginfocom.swing.MigLayout
import org.gjt.jclasslib.browser.BrowserFrame
import org.gjt.jclasslib.util.DefaultAction
import org.gjt.jclasslib.util.GUIHelper
import org.gjt.jclasslib.util.ProgressDialog
import java.awt.event.*
import javax.swing.*
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreePath

class ClasspathBrowser(private val frame: BrowserFrame, title: String, private val updateClassPathFromFrame: Boolean) : JDialog(frame, title) {

    var classpathComponent: ClasspathComponent? = null
        set(classpathComponent) {
            if (field != classpathComponent) {
                resetOnNextMerge = field.let { it == null || classpathComponent?.contains(it) == false }
                needsMerge = true
                field = classpathComponent
                if (resetOnNextMerge) {
                    clear()
                }
            }
        }

    private val classPathTree: JTree = createTree("Class Path")
    private val modulePathTree: JTree = createTree("Module Path")
    private val trees = listOf(classPathTree, modulePathTree)

    private val tabbedPane = JTabbedPane().apply {
        for (tree in trees) {
            addTab(tree.name, JScrollPane(tree))
        }
    }

    private fun createTree(treeName : String): JTree {
        return JTree(ClassTreeNode()).apply {
            name = treeName
            isRootVisible = false
            showsRootHandles = true
            putClientProperty("JTree.lineStyle", "Angled")

            addTreeSelectionListener {
                val selectionPath = selectionPath
                okAction.isEnabled = if (selectionPath != null) {
                    !(selectionPath.lastPathComponent as ClassTreeNode).isPackageNode
                } else false

            }
            addMouseListener(object : MouseAdapter() {
                override fun mouseClicked(event: MouseEvent) {
                    if (event.clickCount == 2 && isValidDoubleClickPath(event, this@apply)) {
                        okAction()
                    }
                }
            })
        }
    }

    private val setupAction = DefaultAction("Setup classpath") {
        frame.setupClasspathAction.actionPerformed(ActionEvent(this, 0, null))
        classpathComponent = frame.config.toImmutableContainer()
        conditionalUpdate()
    }

    private val syncAction = DefaultAction("Synchronize") {
        sync(true)
    }

    private val okAction = DefaultAction("OK") {
        isVisible = false
    }.apply {
        isEnabled = false
    }

    var isCanceled: Boolean = false
        private set

    private val cancelAction = DefaultAction("Cancel") {
        isVisible = false
        isCanceled = true
    }.apply {
        accelerator(KeyEvent.VK_ESCAPE, 0)
        applyAcceleratorTo(contentPane as JComponent)
    }

    private val progressDialog: ProgressDialog by lazy {
        ProgressDialog(this, "Scanning classpath ...")
    }

    private var resetOnNextMerge: Boolean = false
    private var needsMerge: Boolean = false

    /**
     * Get the name of the selected class.
     * @return the name
     */
    val selectedClassNames: Collection<String>
        get() {
            val selectionPaths = getSelectedTree().selectionPaths?.toList() ?: emptyList()
            return selectionPaths.map {selectionPath ->
                val buffer = StringBuilder()
                for (i in 1 until selectionPath.pathCount) {
                    if (buffer.isNotEmpty()) {
                        buffer.append('/')
                    }
                    buffer.append(selectionPath.getPathComponent(i).toString())
                }
                buffer.toString()
            }
        }

    init {
        setupComponent()
    }

    fun isModulePathSelection(): Boolean = getSelectedTree() == modulePathTree

    override fun setVisible(visible: Boolean) {
        if (visible) {
            isCanceled = false
            if (updateClassPathFromFrame) {
                classpathComponent = frame.config.toImmutableContainer()
            }
        }
        super.setVisible(visible)
    }

    fun clear() {
        //The trees will not be synchronized automatically on the next setVisible.
        for (tree in trees) {
            tree.model = DefaultTreeModel(ClassTreeNode())
        }
    }

    private fun setupComponent() {
        (contentPane as JComponent).apply {
            layout = MigLayout("wrap", "[grow]", "[grow]" + (if (updateClassPathFromFrame) "[]" else "") + "para[nogrid]")

            add(tabbedPane, "grow")

            if (updateClassPathFromFrame) {
                add(setupAction.createTextButton(), "tag help2")
            }
            add(syncAction.createTextButton(), "tag help2")
            add(okAction.createTextButton().apply {
                this@ClasspathBrowser.getRootPane().defaultButton = this
            }, "tag ok")
            add(cancelAction.createTextButton(), "tag cancel")

            addWindowListener(object : WindowAdapter() {
                override fun windowClosing(event: WindowEvent?) {
                    cancelAction()
                }

                override fun windowActivated(e: WindowEvent?) {
                    conditionalUpdate()
                }
            })
        }

        setSize(450, 450)
        isModal = true
        GUIHelper.centerOnParentWindow(this, owner)
        defaultCloseOperation = WindowConstants.DO_NOTHING_ON_CLOSE
    }

    private fun conditionalUpdate() {
        if (needsMerge) {
            sync(resetOnNextMerge)
        }
    }

    private fun isValidDoubleClickPath(event: MouseEvent, tree: JTree): Boolean {
        val locationPath = tree.getPathForLocation(event.x, event.y)
        val selectionPath = tree.selectionPath
        if (selectionPath == null || locationPath == null || selectionPath != locationPath) {
            return false
        }
        return !(selectionPath.lastPathComponent as ClassTreeNode).isPackageNode
    }

    private fun sync(reset: Boolean) {
        val classPathModel = getModel(reset, classPathTree)
        val modulePathModel = getModel(reset, modulePathTree)
        resetOnNextMerge = false
        needsMerge = false

        progressDialog.task = {
            classpathComponent?.mergeClassesIntoTree(classPathModel, modulePathModel, reset)
        }
        progressDialog.isVisible = true
        if (reset) {
            classPathTree.applyModel(classPathModel)
            modulePathTree.applyModel(modulePathModel)
        }
    }

    private fun JTree.applyModel(model: DefaultTreeModel) {
        this.model = model
        expandPath(TreePath(model.root))
    }

    private fun getModel(reset: Boolean, tree: JTree) = if (reset) DefaultTreeModel(ClassTreeNode()) else tree.model as DefaultTreeModel

    private fun getSelectedTree(): JTree = trees[tabbedPane.selectedIndex]
}
